javascript/ql/src/semmle/javascript/dataflow/Configuration.qll

```diff
@@ -302,42 +308,27 @@ abstract class BarrierGuardNode extends DataFlow::Node {
     exists(SsaRefinementNode ref, boolean outcome |
       nd = DataFlow::ssaDefinitionNode(ref) and
       forex(SsaVariable input | input = ref.getAnInput() |
-        asExpr() = ref.getGuard().getTest() and
+        getExpr() = ref.getGuard().getTest() and
         outcome = ref.getGuard().(ConditionGuardNode).getOutcome() and
-        internalBlocksExpr(outcome, input.getAUse(), label)
+        barrierGuardBlocksExpr(this, outcome, input.getAUse(), label)
       )
     )
     or
     // 2) `nd` is an instance of an access path `p`, and dominated by a barrier for `p`
     exists(AccessPath p, BasicBlock bb, ConditionGuardNode cond, boolean outcome |
       nd = DataFlow::valueNode(p.getAnInstanceIn(bb)) and
-      asExpr() = cond.getTest() and
+      getExpr() = cond.getTest() and
       outcome = cond.getOutcome() and
-      internalBlocksAccessPath(outcome, p, label) and
+      barrierGuardBlocksAccessPath(this, outcome, p, label) and
       cond.dominates(bb)
     )
   }
 
-  /**
-   * Holds if data flow node `nd` acts as a barrier for data flow.
-   *
-   * `label` is bound to the blocked label, or the empty string if all labels should be blocked.
-   */
-  private predicate internalBlocksExpr(boolean outcome, Expr test, string label) {
-    blocks(outcome, test) and label = ""
+  /** Gets the corresponding expression, including that of reflective calls. */
+  private Expr getExpr() {
+    result = asExpr()
     or
-    blocks(outcome, test, label)
-  }
-
-  /**
-   * Holds if data flow node `nd` acts as a barrier for data flow due to aliasing through
-   * an access path.
-   *
-   * `label` is bound to the blocked label, or the empty string if all labels should be blocked.
-   */
-  pragma[noinline]
-  private predicate internalBlocksAccessPath(boolean outcome, AccessPath ap, string label) {
-    internalBlocksExpr(outcome, ap.getAnInstance(), label)
+    this = DataFlow::reflectiveCallNode(result)
```

Comment:
I think I've previously had the same issue as the one you are solving here. 

Could there be an idea in changing `DataFlow::Node::asExpr()` to include reflective calls, or lift this method to `DataFlow::Node`?

Reply:
In any case `getExpr` might not be the most descriptive name.  

`asExprInclReflectiveCalls`? (But that name feels too long, feel free to ignore).

Reply:
> Could there be an idea in changing DataFlow::Node::asExpr() 

I think that leads to a recursion mess, and also a mess of contepts. I would prefer the second option:

>  lift this method to DataFlow::Node

But a good name escapes me.  `asUnderlyingExpr` is my best suggestion.

Reply:
My main gripe with using `asExpr()` is that two data flow nodes can suddenly map to the same expression, which brings along its own set of new pitfalls. But maybe we should do it anyway? It's essentially the same issue as parameter nodes not having AST nodes.

The general solution is to take the thing you know is an expression (`getTest()`), map that to a data-flow node, and follow up with `getImmediatePredecessor`, `getAPredecessor`, or `getALocalSource`. The third version I mentioned in the PR description used `getImmediatePredecessor` but I backed it out again, because it required more work for that generalization to actually work beyond reflective calls.

> I think I've previously had the same issue as the one you are solving here.

@erik-krogh  Can you find out what it was? It might help us figure out which solution works best.

Reply:
> @erik-krogh Can you find out what it was? It might help us figure out which solution works best.

From looking around I can't immediately find an instance of me solving the problem, but I think that is because I ended up not solving it.  
The problem definitely exists in `UseOfReturnlessFunction`, which has no explicit handling of reflective calls, and `DataFlow::CallNode::asExpr()` is therefore `none()`. 

As a consequence, `UseOfReturnlessFunction` has an FP on the last line in this example: 
```JavaScript
function onlySideEffects() {
  console.log("Boo!")
}
var dead1 = onlySideEffects(); // Correctly not flagged
var dead2 = onlySideEffects.call(null); // FP! flagged by UseOfReturnlessFunction
```

Reply:
I'm currently doing a dist-compare on an implementation of `DataFlow::Node::getEnclosingExpr()`. 
```
Expr getEnclosingExpr() {
  result = asExpr() or
  this = DataFlow::reflectiveCallNode(result) or
  result = this.(ParameterNode).getParameter()
}
```

The evaluation is on `UseOfReturnlessFunction`, where i changed some `asExpr()` to `getEnclosingExpr()`, which definitely removes some false positives. 
If the results are good I'll land it in a PR.